plugins/zstack.py

```diff
@@ -0,0 +1,372 @@
+        # Mvoe back to start
+        logging.debug("Z Stack acquisition complete. Returning focus to start position")
+        self.focus.moveAbs(self.old_pos).result()
+        
+    def constructCube(self, images):
+        # images is a list of 5 dim data arrays.
+        ret = []
+        for image in images:
+            stack = np.dstack(image)
+            #stack = np.swapaxes(stack, 0, 2)
+            #stack = np.expand_dims(stack, 1)
+            ret.append(stack[0])
+            
+        # Add back metadata
+        metadata3d = copy.copy(self._metadata)
+        # Extend pixel size to 3D
+        ps_x, ps_y = metadata3d[model.MD_PIXEL_SIZE]
+        ps_z = self.zstep.value
+        metadata3d[model.MD_PIXEL_SIZE] = (ps_x, ps_y, ps_z)
+        metadata3d[model.MD_DIMS] = "ZYX"
+        
+        ret = DataArray(ret, metadata3d)
+            
+        return ret
+
+    # ALL GENERIC
+    def acquire(self, dlg):
+        main_data = self.main_app.main_data
+        str_ctrl = main_data.tab.value.streambar_controller
+        stream_paused = str_ctrl.pauseStreams()
+
+        nb = self.numberofAcquisitions.value
+        ss, last_ss = self._get_acq_streams()
+
+        sacqt = acq.estimateTime(ss)
+        speed = self.focus.speed.value['z']
+        step_time = driver.estimateMoveDuration(abs(self.zstep.value), speed, 0.01)
+
+        logging.info("Z stack acquisition started with %d levels with streams %s", nb, ss)
+
+        # Specific plugin init
+        self.initAcquisition()
+
+        # TODO: if drift correction, use it over all the time
+        f = model.ProgressiveFuture()
+        f.task_canceller = lambda l: True  # To allow cancelling while it's running
+        f.set_running_or_notify_cancel()  # Indicate the work is starting now
+        dlg.showProgress(f)
+        
+        images = None
+
+        for i in range(nb):
+            left = nb - i
+            dur = sacqt * left + step_time * (left - 1)
+            
+            logging.debug("Z Stack Acquisition %d of %d", i, nb)
+            
+            if left == 1 and last_ss:
+                ss += last_ss
+                dur += acq.estimateTime(ss) - sacqt
+
+            startt = time.time()
+            f.set_progress(end=startt + dur)
+            das, e = acq.acquire(ss).result()
+            if images is None:
+                # Copy metadata from the first acquisition
+                self._metadata = copy.copy(das[0].metadata)
+                images = [[] for i in range(len(das))]
+            
+            for im, da in zip(images, das):
+                im.append(da)
+
+            if f.cancelled():
+                return
+
+            # Execute an action to prepare the next acquisition for the ith acquisition
+            self.stepAcquisition(i)
+
+        f.set_result(None)  # Indicate it's over
+        
+        # Construct a cube from each stream's image. 
+        cubes = [self.constructCube(im) for im in images]
```

Comment:
That is not very "generic". You should just have one function `completeAcquisition(das)` which takes all the data received during the acquisition, and return the data as it should be stored. Then the part saving to a file can be generic. It can also take care of extra movements needed at the end, which is what you currently do in `completeAcquisition()`.